### PR TITLE
feat(symcache): Add usym transformer for dSYMs

### PR DESCRIPTION
**Pull request by relaxolotl**
This is a first shot attempt at adding a usym transformer for dSYMs. This is an alternative to #540 which requires both a dSYM and a usym to symbolicate IL2CPP-compiled Unity projects, all the way up to their managed symbols (i.e. C# code). 

This basically slaps all the additional lookup functionality needed by the transformer to `UsymSymbols` and calls it a day. I've tested this against the dSYMs and usyms generated by the sentry test unity il2cpp etc project, but I have no idea how to trim down the dSYM in a safe and non-tedious way so that it isn't an 90-something MB file, making it a bit of a chore to create tests for. Ideas are welcome.

### CODE REVIEW

File: symbolic-il2cpp/src/usym.rs [hunk 1]
```diff
@@ -468,10 +468,44 @@ impl<'a> UsymSymbols<'a> {
...
+            }) {
+            Ok(index) => self.get_record(index),
+            Err(0) => None,
+            Err(index) => self.get_record(index - 1),
+        }
+    }
+
+    /// Look up the managed code source location for some given native symbol.
+    pub fn lookup_by_symbol(&self, symbol: String) -> Option<UsymSourceRecord> {
+        match self.records.binary_search_by_key(&symbol, |candidate| {
```
Comment (Swatinem) [inline symbolic-il2cpp/src/usym.rs:502]:
```suggestion
    pub fn lookup_by_symbol(&self, symbol: &str) -> Option<UsymSourceRecord> {
        match self.records.binary_search_by_key(symbol, |candidate| {
```

That way you don’t need the `to_string` in the transformer.

File: symbolic-il2cpp/src/usym.rs [hunk 2]
```diff
@@ -468,10 +468,44 @@ impl<'a> UsymSymbols<'a> {
...
+            .binary_search_by_key(&instruction_pointer, |r| r.address)
+        {
+            Ok(index) => self.get_record(index),
+            Err(0) => None,
+            Err(index) => self.get_record(index - 1),
+        }
+    }
+
+    /// Look up the managed code source location for some filename and line.
+    pub fn lookup_by_srcloc(&self, file: String, line: u32) -> Option<UsymSourceRecord> {
```
Comment (Swatinem) [inline symbolic-il2cpp/src/usym.rs:485]: similar here:

```suggestion
    pub fn lookup_by_srcloc(&self, file: &str, line: u32) -> Option<UsymSourceRecord> {
```

You are calling `into_owned()` for each step of the search, which is super expensive.

**Comment by relaxolotl** [comment]
I've decided to take a different approach instead: addresses are injected into `transform::Function` and `transform::SourceLocation` so they can be used to look up their respective records in the usym. 

The records in usyms are sorted by address, and to look them up by any other field would mean one of two things: we'd need to use a different search algorithm that doesn't rely on the records already being sorted to begin with, or we'd have to sort the records by the other lookup fields and cache them. 

Seeing as we're already not big fans of string allocation based on the existing comments on the PR, I would imagine the caching idea would be heavily rejected as we'd have to store at minimum 2 separate copies of the usym records: one sorted by the key used to look up functions and the other by the key used for source locations. Trying to find or derive a common key between those two structs so we could use one common sorted list of records points us towards using addresses, as they're the only common piece of data available across functions and source locations (line info, to be specific). This is also why addresses are being used as the lookup key here.

The new, address-based approach comes with one flaw: Sometimes, this is unable to fetch the managed name for functions. This is because those functions have an `entry_pc` which points to an unmapped record. I'm not sure how or why this happens. Let's use the il2cpp line number unity project as an example: `NewBehaviourScript.Start()` has an `entry_pc` of `0x7df180`. It has two lines:
```
[
        LineInfo {
            address: 0x7df180,
            size: 0x8,
            file: FileInfo {
                name: "Assembly-CSharp.cpp",
                dir: "Classes/Native",
            },
            line: 145,
        },
        LineInfo {
            address: 0x7df188,
            size: 0x4,
            file: FileInfo {
                name: "Assembly-CSharp.cpp",
                dir: "Classes/Native",
            },
            line: 149,
        },
    ]
```

The second line has a mapped record:
```
MappedRecord {
    address: 0x7df188,
    native_symbol: "NewBehaviourScript_Start_mDEAA416D3A6676A79851DDF6B62D676E53372A87",
    native_file: "/Users/bitfox/_Workspace/IL2CPP/Builds/iOS/Xcode/Classes/Native/Assembly-CSharp.cpp",
    native_line: 149,
    managed_symbol: "NewBehaviourScript.Start()",
    managed_file_info: FileInfo {
        name: "NewBehaviourScript.cs",
        dir: "/Users/bitfox/_Workspace/IL2CPP/Assets",
    },
    managed_line: 10,
}
```
Its native line can be found here: https://github.com/getsentry/sentry-unity-il2cpp-line-numbers/blob/main/Builds/iOS/Xcode/Classes/Native/Assembly-CSharp.cpp#L149.

The first line, which determines the function's `entry_pc` points to an unmapped record:
```
UnmappedRecord {
    address: 0x7df180,
    native_symbol: "NewBehaviourScript_Start_mDEAA416D3A6676A79851DDF6B62D676E53372A87",
    native_file: "/Users/bitfox/_Workspace/IL2CPP/Builds/iOS/Xcode/Classes/Native/Assembly-CSharp.cpp",
    native_line: 145,
}
```
Which refers to this native line: 
Its native line can be found here: https://github.com/getsentry/sentry-unity-il2cpp-line-numbers/blob/main/Builds/iOS/Xcode/Classes/Native/Assembly-CSharp.cpp#L145.

On the other hand, `NewBehaviourScript.Update()` doesn't have this problem, and its `entry_pc` points to a mapped function. I'm not sure what the difference between the two is here.

**Comment by loewenheim** [comment]
That this (a function's entry pc not pointing to managed code) happens is not too surprising, I guess—maybe the function just starts with some native bookkeeping stuff that doesn't directly correspond to anything in the managed code. What to do about it is another question. One somewhat kludgy solution I can imagine is to get the record with the function's entry pc, regardless of whether it's a mapped or unmapped one (if it doesn't exist, I think we're out of luck). From there, scan forward linearly through records that refer to the same native function until you hopefully find a mapped one.

**Comment by relaxolotl** [comment]
> One somewhat kludgy solution I can imagine is to get the record with the function's entry pc, regardless of whether it's a mapped or unmapped one (if it doesn't exist, I think we're out of luck). From there, scan forward linearly through records that refer to the same native function until you hopefully find a mapped one.

This did cross my mind, but I didn't spend too much time thinking about it but hey, let me take a crack at exploring the idea in detail right now. With the forward-scanning idea, we'd probably have to introduce a separate address-lookup function, as the forward-scanning behaviour isn't needed for source locations. We've got a few scenarios to handle:

1. Mapped record is found immediately for the address
   1. It's a direct hit/Exact address match (trivial/happy path, can ignore)
   1. It's a best match hit with a record for the closest previous address
      1. Record belongs to the function (trivial/happy path, can ignore)
      1. Record doesn't belong to the function (this is actually a bug in the PR right now)
         1. There is a record that belongs to the function if you continue scanning forward
         1. There aren't any more records that belong to the function if you continue scanning forward
1. An unmapped record is found for the address
   1. Scan forward until the first mapped function is found
      1. Record belongs to the function
      1. Record does not belong to the function
         1. There is a record that belongs to the function if you continue scanning forward
         1. There aren't any more records that belong to the function if you continue scanning forward
   1. Scan forward but there are no more mapped functions following the direct hit
1. No records can be found for the address

It would be ideal if a function could actually pass in a list of addresses belonging to all of the lines assigned to it so we could constrain the amount of times we seek down <N (N=# of records in usym) times. This should be doable. What would cause issues if we use this assumption is if a mapped record we're looking for isn't referenced by any lines in the function extracted from the dSYM, but the record _does_ live in the usym somewhere. I'm skeptical that this is a problem, but I could be completely wrong about this. For now though, I'm just going to pretend this isn't a problem.

We now need to figure out what should take care of the responsibility of determining whether a best guess record actually lines up with the function (ie who's going to take care of comparing the native symbols from the dSYM and the best guess record?) If we can assume that the addresses for all lines in a function are sorted and weakly increasing, we can thread the native symbol name through the transformer down into the lookup function on usym, so it can just seek forwards until it finds a matching mapped record. Alternatively, we can pass in a closure that tells the lookup function whether some given "backup" record matches what its invoker is looking for.

This leaves us with the following algorithm: 
1. Pass a list of addresses belonging to the function into the transformer.
1. Look up the address, and double check that the native symbols line up.
1. If the native symbols line up, return the updated function.
1. If the native symbols don't line up, grab the next address and go back to 2.
1. If the native symbols don't line up and there aren't any addresses left, return the original function.

Thoughts?

**Comment by Swatinem** [comment]
We are not going forward with usym, but I will keep this branch alive for future reference.

**Pull request closed without merging**